enterprise/backend/src/metabase_enterprise/sandbox/models/params/field_values.clj

```diff
@@ -0,0 +1,87 @@
+(ns metabase-enterprise.sandbox.models.params.field-values
+  (:require [clojure.core.memoize :as memoize]
+            [metabase-enterprise.enhancements.ee-strategy-impl :as ee-strategy-impl]
+            [metabase-enterprise.sandbox.api.table :as sandbox.api.table]
+            [metabase.api.common :as api]
+            [metabase.models.field :as field :refer [Field]]
+            [metabase.models.field-values :as field-values :refer [FieldValues]]
+            [metabase.models.params.field-values :as params.field-values]
+            [metabase.public-settings.metastore :as settings.metastore]
+            [metabase.util :as u]
+            [pretty.core :as pretty]
+            [toucan.db :as db]
+            [toucan.hydrate :refer [hydrate]]))
+
+(comment api/keep-me)
+
+(def ^:private ^{:arglist '([last-updated field])} fetch-sandboxed-field-values*
+  (memoize/ttl
+   ;; use a custom key fn for memoization. The custom key includes current User ID and a hash of their permissions
+   ;; set, so we cache per-User (and so changes to that User's permissions will result in a cache miss), and Field ID
+   ;; instead of an entire Field object (so maps with slightly different keys are still considered equal)
+   ^{::memoize/args-fn (fn [[updated-at field]]
+                         [api/*current-user-id*
+                          (hash @api/*current-user-permissions-set*)
+                          updated-at
+                          (u/the-id field)])}
+   (fn [_ field]
+     {:values   (field-values/distinct-values field)
+      :field_id (u/the-id field)})
+   ;; TODO -- shouldn't we return sandboxed human-readable values as well??
+   ;;
+   ;; Expire entires older than 30 days so we don't have entries for users and/or fields that
+   ;; no longer exists hanging around.
+   ;; (`clojure.core.cache/TTLCacheQ` (which `memoize` uses underneath) evicts all stale entries on
+   ;; every cache miss)
+   :ttl/threshold (* 1000 60 60 24 30)))
+
+(defn- fetch-sandboxed-field-values
+  [field]
+  (fetch-sandboxed-field-values*
+   (db/select-one-field :updated_at FieldValues :field_id (u/the-id field))
```

Comment:
The timestamp here is used as cache invalidation when there would be new field values. If a field has no FieldValues (too many distinct values, temporal, marked manually as don't cache them) this will be nil and there would be no way to get newer field values afterwards. This is certainly an edge case on fields that get promoted out of list or auto-list. Is this endpoint ever hit for other types of fields, where it doesn't expect to find FieldValues?

Reply:
I don't think so. Fields have a hydrated `has_field_values` key we add in API responses and I think the FE doesn't hit this endpoint in that case